RO-2037: Offline registrations using sqlite

jorgkv opened the pull request:
Jira-sak: https://nveprojects.atlassian.net/browse/RO-2037

Viktigste endringer:
- Legger til en ny avhengighet i appen for bruk av sqlite native, og en ny sqlite-service for håndtering av observasjoner offline
- Legger til en ny service som kan override vanlig online-søk i appen, `OfflineCapableSearchService`.

Det gjenstår fortsatt arbeid knytta til offline-observasjoner i appen, dette ligger i søk- og filter-epic taget med offline-observasjoner. Viktigst er nok å få til at slettede observasjoner også slettes fra offline-observasjonene i appen, det gjøres ikke i dag.

# Må løses før denne tas inn:

- [x] Sjekk inn eventuelle endringer i android-mappa (@gruble eller @amish1188)
Øystein: Så ikke ut som det ble noen endringer i denne mappa etter bygging.

jorgkv committed:
Add search service with offline caching of registrations

jorgkv commented:
Fikk appen til å kræsje med hvit skjerm som ikke klarte å starte appen igjen, dette ble logget:
```
2022-11-28 12:58:01.987781+0100 App[15590:4864799] [Process] 0x10d001b80 - [PID=15596] WebProcessProxy::didClose: (web process 0 crash)
2022-11-28 12:58:01.987868+0100 App[15590:4864799] [Process] 0x10d001b80 - [PID=15596] WebProcessProxy::processDidTerminateOrFailedToLaunch: reason=Crash
2022-11-28 12:58:01.987917+0100 App[15590:4864799] [ProcessSuspension] 0x10c020d80 - ProcessAssertion: Failed to acquire RBS Background assertion 'ConnectionTerminationWatchdog' for process because PID 0 is invalid
2022-11-28 12:58:01.987981+0100 App[15590:4864799] [Process] 0x13a02e018 - [pageProxyID=7, webPageID=8, PID=15596] WebPageProxy::processDidTerminate: (pid 15596), reason=Crash
2022-11-28 12:58:01.990516+0100 App[15590:4867521] [ProcessSuspension] 0x10c020d80 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'ConnectionTerminationWatchdog' for process with PID=0, error: (null)
2022-11-28 12:58:01.993248+0100 App[15590:4864799] [Loading] 0x13a02e018 - [pageProxyID=7, webPageID=8, PID=15596] WebPageProxy::dispatchProcessDidTerminate: reason=Crash
2022-11-28 12:58:02.015336+0100 App[15590:4867581] [assertion] Error acquiring assertion: <Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}>
2022-11-28 12:58:02.015416+0100 App[15590:4867581] [ProcessSuspension] 0x10c021ec0 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'WebProcess Background Assertion' for process with PID=15596, error: Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}
Error Domain=WKErrorDomain Code=5 "JavaScript-utføringen returnerte et resultat av en type som ikke støttes" UserInfo={NSLocalizedDescription=JavaScript-utføringen returnerte et resultat av en type som ikke støttes}
2022-11-28 12:58:02.137676+0100 App[15590:4864799] [Process] 0x10c06c640 - GPUProcessProxy::gpuProcessExited: reason=IdleExit
2022-11-28 12:58:02.137722+0100 App[15590:4864799] [Process] 0x10d003510 - [PID=15600] WebProcessProxy::gpuProcessExited: reason=IdleExit
2022-11-28 12:58:02.139722+0100 App[15590:4867521] [assertion] Error acquiring assertion: <Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}>
2022-11-28 12:58:02.139749+0100 App[15590:4867521] [ProcessSuspension] 0x10c020180 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'GPUProcess Background Assertion' for process with PID=15593, error: Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}
2022-11-28 12:58:31.989066+0100 App[15590:4864799] Could not signal service com.apple.WebKit.WebContent: 113: Could not find specified service
```

jorgkv commented:
Litt urovekkende at appen kræsjer såpass ofte på ios. Jeg har ikke sjekket om det er værre enn med prod-versjon, men det kræsjer ofte. Minnebruk ligger på ca 110mb hele tiden uten at det øker nevneverdig før det kræsjer. Samme med CPU som ligger på < 5% hele tiden. Så jeg vet ikke helt hva det kan skyldes.

jorgkv commented:
<details>
<summary>Fikk appen til å kærsje også i develop med en lignende feilmelding</summary>
<pre>
2022-11-28 13:21:32.233170+0100 App[15641:4877768] [Process] 0x10f0009d0 - [PID=15642] WebProcessProxy::didClose: (web process 0 crash)
2022-11-28 13:21:32.233254+0100 App[15641:4877768] [Process] 0x10f0009d0 - [PID=15642] WebProcessProxy::processDidTerminateOrFailedToLaunch: reason=Crash
2022-11-28 13:21:32.233332+0100 App[15641:4877768] [ProcessSuspension] 0x10e020300 - ProcessAssertion: Failed to acquire RBS Background assertion 'ConnectionTerminationWatchdog' for process because PID 0 is invalid
2022-11-28 13:21:32.233430+0100 App[15641:4877768] [Process] 0x138032018 - [pageProxyID=7, webPageID=8, PID=15642] WebPageProxy::processDidTerminate: (pid 15642), reason=Crash
2022-11-28 13:21:32.238454+0100 App[15641:4879782] [ProcessSuspension] 0x10e020300 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'ConnectionTerminationWatchdog' for process with PID=0, error: (null)
2022-11-28 13:21:32.252289+0100 App[15641:4879782] [assertion] Error acquiring assertion: <Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}>
2022-11-28 13:21:32.252445+0100 App[15641:4879782] [ProcessSuspension] 0x10e020360 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'WebProcess Background Assertion' for process with PID=15642, error: Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}
2022-11-28 13:21:32.265375+0100 App[15641:4877768] [Loading] 0x138032018 - [pageProxyID=7, webPageID=8, PID=15642] WebPageProxy::dispatchProcessDidTerminate: reason=Crash
2022-11-28 13:21:32.394035+0100 App[15641:4877768] [Process] 0x10e06c4e0 - GPUProcessProxy::didClose:
2022-11-28 13:21:32.394072+0100 App[15641:4877768] [Process] 0x10e06c4e0 - GPUProcessProxy::gpuProcessExited: reason=Crash
2022-11-28 13:21:32.394094+0100 App[15641:4877768] [Process] 0x10f0030a0 - [PID=15648] WebProcessProxy::gpuProcessExited: reason=Crash
2022-11-28 13:21:32.395679+0100 App[15641:4879387] [assertion] Error acquiring assertion: <Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}>
2022-11-28 13:21:32.395708+0100 App[15641:4879387] [ProcessSuspension] 0x10e020720 - ProcessAssertion::acquireSync Failed to acquire RBS assertion 'GPUProcess Background Assertion' for process with PID=15644, error: Error Domain=RBSServiceErrorDomain Code=1 "target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit" UserInfo={NSLocalizedFailureReason=target is not running or doesn't have entitlement com.apple.runningboard.assertions.webkit}
</pre>
</details>

gruble reviewed (commented):
Jeg får ikke se noen observasjoner når jeg velger testmiljø. Har en mistanke om at jeg jeg brukte testmiljø når jeg ikke hadde wifi-forbindelse til API'et, og at den satte sist-oppdatert-tidspunktet selv om den ikke fant noen obser.
Hver gang jeg søker på nytt nå finner den ingenting, og det er kanskje ikke så rart med disse kriteriene:
FromDtChangeTime: "2022-11-28T14:25:17+01:00"
FromDtObsTime: "2022-11-14T14:25:19+01:00"
LangKey: 1
NumberOfRecords: 500
Offset: 0

jorgkv commented:
> Jeg får ikke se noen observasjoner når jeg velger testmiljø. Har en mistanke om at jeg jeg brukte testmiljø når jeg ikke hadde wifi-forbindelse til API'et, og at den satte sist-oppdatert-tidspunktet selv om den ikke fant noen obser. Hver gang jeg søker på nytt nå finner den ingenting, og det er kanskje ikke så rart med disse kriteriene: FromDtChangeTime: "2022-11-28T14:25:17+01:00" FromDtObsTime: "2022-11-14T14:25:19+01:00" LangKey: 1 NumberOfRecords: 500 Offset: 0

Ok, jeg vet ikke hvordan det kan ha skjedd, gjerne sjekk om du skjønner hvorfor.

Hvis vi skal gå videre med dette og det viser seg å være et problem at ting kommer ut av sync så bør vi kanskje jevnlig kjøre en count på antall observasjoner siste to uker og sjekke om det stemmer med det vi har basen.

Ø: Ser ut som jeg fikk dem inn likevel etterhvert. Vet ikke hva som gjorde at den plutselig fikk orden i sysakene. Skal titte litt i logger

gruble commented:
Fant noe annet rart nå. Virker som appen ikke får helt med seg at jeg bytter naturfare.
Isobser vises fortsatt i kartet samme hvilken naturfare jeg velger. Ingen ny spørring mot databasen logges, annet enn synkroniseringsjobben.
![image](https://user-images.githubusercontent.com/71138449/204297105-8800add0-cdf4-4f47-a9f0-c513f0a41978.png)

jorgkv committed:
Force usage of offline service

  - NB: Remove before PR finish

gruble commented on src/app/app.component.ts:
```diff
@@ -118,8 +120,15 @@ export class AppComponent {
-        )
+          ];
+
+          // if (isPlatform('hybrid')) {
```
Regner med at platformsjekken her er midlertidig kommentert ut for testing, og at vi kun skal bruke denne i app?

gruble commented on src/app/core/services/search-registration/offline-capable-search-service.ts:
```diff
@@ -1,27 +1,287 @@
+const SYNC_INTERVAL = 120000;
+
+let _syncId = 0;
+const getSyncId = (appMode: AppMode, lang: LangKey) => {
```
skal vi kalle denne createSyncId() i stedet? Siden den lager en ny id hver gang?

jorgkv replied:
Fiksa

amish1188 committed:
bring back resetPaging (#442)

gruble commented on src/app/core/services/search-registration/offline-capable-search-service.ts:
```diff
@@ -1,27 +1,287 @@
+type SyncRequestResult = 'success' | 'error';
+const SYNC_DEBOUNCE_MS = 500;
+const SYNC_TIMEOUT_MS = 5000; // Includes the debounce time
+const SYNC_INTERVAL = 120000;
```
Jeg synes dette høres litt ofte ut, men hvis synkronisering er billig er det nok greit;)

jorgkv replied:
Jeg tror 2 minutter er ganske mye, tror ikke det er mange som har appen åpen så lenge.

gruble commented on src/app/core/services/search-registration/offline-capable-search-service.ts:
```diff
@@ -1,27 +1,287 @@
+};
 
 /**
  * Will search for observations in the offline database instead of searching online through the Regobs API
```
Den håndterer vel også synk / henting av observasjoner fra API både på tid og forespørsel? Tenker det er verdt å nevne

jorgkv replied:
La til en kommentar som forklarer dette bedre ja.

gruble commented on src/app/core/services/search-registration/offline-capable-search-service.ts:
```diff
@@ -1,27 +1,287 @@
 @Injectable()
 export class OfflineCapableSearchService extends SearchService {
-  constructor(config: RegobsApiConfiguration, http: HttpClient, private logger: LoggingService) {
+  private syncRequests = new BehaviorSubject<Subject<SyncRequestResult>[]>([]);
```
Jeg forstår ikke hvorfor syncRequests trenger en Subject inni en annen Subject?

jorgkv replied:
Har lagt til en kommentar som forklarer dette bedre nå, og gjort om `Subject<SyncRequestResult>[]` til en egen type.

gruble reviewed (commented):
Det går kjapt å hente observasjoner fra lokal database og virker som filter på kartutsnitt fungerer utmerket i listevisning.

Når jeg bytter språk er det litt som skjer:

1. Jeg får denne i konsollet:
TypeError: Cannot read properties of undefined (reading 'sync_time_ms') at sqlite.service.ts:225:29

2. Observasjonsmarkører i kartet forsvinner, selv om det ser ut som den henter observasjoner på nytt fra API'et og lagrer dem i databasen. Markørene vises ikke når den er ferdig med synk.

3. Det tar rundt to minutter å hente og lagre alle observasjoner på nytt når man bytter språk (i prod nå). Mulig det tar mye tid pga. mye logging, men det kan kanskje være verdt å prøve å kjøre insert av flere observasjoner om gangen.

gruble reviewed (changes requested):
Hvis jeg er i listevisning og har scrollet nedover et lite stykke, scroller den ikke til toppen når jeg bytter naturfare. Det virker som den henter riktige observasjoner, men den scroller ikke til toppen. På beta.regobs.no gjør den dette.

jorgkv committed:
Document what syncRequests is

jorgkv committed:
Rename getSyncId to createSyncId

jorgkv committed:
Improve offline service docstring

jorgkv committed:
Document pagedSearch better

jorgkv committed:
Insert all new registrations and not one at a time

jorgkv committed:
Remove bloated logging params

jorgkv committed:
Try to make the sync operation faster

It now sends the next request to the api
before saving registrations to the db.

jorgkv committed:
Log fail to read last sync ms as a warning

jorgkv committed:
Temp commit

jorgkv committed:
Show toast if observations are outdated

jorgkv committed:
Remove unused spec file

jorgkv commented:
Denne bør squashes inn, det er feks en temp commit jeg ikke har rukket å rydde opp i.

jorgkv commented:
Anser denne som ferdig nok til at vi kan vurdere å ta den inn i prosjektet, men det gjenstår feks å sjekke inn android-filer, om det er noen. I tillegg ligger resterende oppgaver i Jira. 

Bør testes:
 - Generell bruk
 - Se gjerne om dere ser en bedre løsning på dette med å sette `lastFetched`-tidspunkt, kanskje litt hackete den løsningen som ligger inne nå
 - Toast for gamle observasjoner, sjekk at den dukker opp ved å starte appen uten nett osv (må være 15 min siden sist hentet observasjoner for at den skal dukke opp)
 - Sjekk at innsendte observasjoner dukker opp i kart og listevisning hvis man går offline rett etter å ha sendt inn obs, før man navigerer til kart / listevisning

gruble commented:
Detaljside for observasjon virker ikke. Viser blank side når jeg trykker på en at-a-glance-obs.

gruble commented on src/app/core/services/sqlite/sqlite.service.ts:
```diff
@@ -0,0 +1,385 @@
+    this.logger.debug('Reading sync time', DEBUG_TAG, { appMode });
+    const result = await this.conn.query(`SELECT * FROM sync_time WHERE app_mode='${appMode}' AND lang=${lang};`);
+    this.logger.debug('Sync time', DEBUG_TAG, result);
+    return result.values[0].sync_time_ms;
```
Fikk denne i konsollet når jeg bytta fra prod til testmiljø.
[WARNING][OfflineCapableSearchService] Failed to read last sync ms
console-logging.service.ts:38 TypeError: Cannot read properties of undefined (reading 'sync_time_ms')
    at sqlite.service.ts:237:29

Jeg tror API'et ikke var tilgjengelig da dette skjedde.

gruble replied:
Sjekket dette igjen nå. Ser ut som det håndteres greit, selv om det ser litt dramatisk ut i konsollet. Vi trenger kanskje ikke å logge det som en feil siden dette er forventet resultat når det er første gang vi synker for et miljø. Vi kan fikse det når vi blir lei av å se feilmeldinga i konsollet.

jorgkv commented:
> Detaljside for observasjon virker ikke. Viser blank side når jeg trykker på en at-a-glance-obs.

Det var rart, fungerer hos meg. Så du om det ble logget noe som gir oss pekepinn på hva det skyldtes?

https://user-images.githubusercontent.com/8558332/220452746-7931c8f0-732d-4db1-9cd6-03065978eb5a.MOV

gruble commented on src/app/core/services/sqlite/sqlite.service.ts:
```diff
@@ -0,0 +1,385 @@
+        concatMap((state) => (state === 'pause' ? this.closeConn() : this.openConn()))
+      )
+      .subscribe();
+    this.platform.pause.subscribe(() => this.pauseResumeEvent.next('pause'));
+    this.platform.resume.subscribe(() => this.pauseResumeEvent.next('resume'));
```
Siden vi uansett behandler pause og resume forskjellig, vil ikke koden bli enklere hvis vi åpner og lukker connection direkte i subscribe fra pause og resume? (i stedet for å sende det via en subject?)

jorgkv replied:
Ved å håndtere begge eventene samlet kan vi sørge for at ikke noe logikk kjøres samtidig, feks at den prøver å åpne en connection mens den stenger den forrige, osv. Vet ikke hvor viktig det er, men nå håndteres det i hvert fall etter hverandre:

https://github.com/NVE/regObs4/blob/5ee14e3f0850637081906fd6239612df3f72e42c/src/app/core/services/sqlite/sqlite.service.ts#L135-L140

gruble commented on src/app/core/services/sqlite/sqlite.service.ts:
```diff
@@ -0,0 +1,385 @@
+};
+
+const DEBUG_TAG = 'OfflineCapableSearchService - Sqlite';
+const DATABASE_NAME = 'regobs-registrations';
```
Vil det være aktuelt å bruke denne databasen til flere ting? Jeg spør fordi navnet er ganske spesifikt.

jorgkv replied:
Jeg endrer navnet til 'regobs-v2'

gruble commented on src/app/core/services/sqlite/sqlite.service.ts:
```diff
@@ -0,0 +1,385 @@
+    statements: [
+      'ALTER TABLE registrations ADD COLUMN app_mode TEXT;',
+
+      `CREATE TABLE IF NOT EXISTS sync_time (
```
Se kommentar om navn på databasen. Hvis vi ønsker å bruke databasen til flere ting, er det kanskje lurt å bruke et mer spesifikt navn på denne tabellen, så vi skjønner at den brukes kun til synk av registreringer?

jorgkv replied:
Fikser 👍

gruble commented on src/app/core/services/sqlite/sqlite.service.ts:
```diff
@@ -0,0 +1,385 @@
+  {
+    toVersion: 1,
+    statements: [
+      `CREATE TABLE IF NOT EXISTS registrations (
```
NB! Snobbe-kommentar:
Tror det er "standard" blant relasjonsdatabase-fetisjister å bruke entallsform på tabellnavn.
https://stackoverflow.com/questions/338156/table-naming-dilemma-singular-vs-plural-names

jorgkv replied:
Fikser 👍

widforss commented:
Kan bekrefte at det ikke går at gå til obs fra at-a-glance på android. Men det er ikke helt konsekvent, en obs kan ikke fungere, så fungerer den plutselig, og då vil den fortsette å fungere. Det merkelige er at observasjoner som ikke går å nå på den måten fortsatt finnes i list-vyen, så dataen er jo der. Det finnes inget interessant i loggen.

gruble commented:
> Kan bekrefte at det ikke går at gå til obs fra at-a-glance på android. Men det er ikke helt konsekvent, en obs kan ikke fungere, så fungerer den plutselig, og då vil den fortsette å fungere. Det merkelige er at observasjoner som ikke går å nå på den måten fortsatt finnes i list-vyen, så dataen er jo der. Det finnes inget interessant i loggen.

Jeg tror jeg fant årsaken. Detaljside for observasjon bruker fortsatt den gamle ObservationService for å laste observasjoner i app. Så de observasjonene er ikke de samme som lagres lokalt i denne PR. Jeg jobber med en fiks nå.

gruble committed:
Fikset detalj-side for observasjon.

Nå henter vi observasjonen fra databasen.

gruble commented:
> Fant noe annet rart nå. Virker som appen ikke får helt med seg at jeg bytter naturfare. Isobser vises fortsatt i kartet samme hvilken naturfare jeg velger. Ingen ny spørring mot databasen logges, annet enn synkroniseringsjobben. ![image](https://user-images.githubusercontent.com/71138449/204297105-8800add0-cdf4-4f47-a9f0-c513f0a41978.png)



> > Kan bekrefte at det ikke går at gå til obs fra at-a-glance på android. Men det er ikke helt konsekvent, en obs kan ikke fungere, så fungerer den plutselig, og då vil den fortsette å fungere. Det merkelige er at observasjoner som ikke går å nå på den måten fortsatt finnes i list-vyen, så dataen er jo der. Det finnes inget interessant i loggen.
> 
> Jeg tror jeg fant årsaken. Detaljside for observasjon bruker fortsatt den gamle ObservationService for å laste observasjoner i app. Så de observasjonene er ikke de samme som lagres lokalt i denne PR. Jeg jobber med en fiks nå.

Fikset i https://github.com/NVE/regObs4/pull/378/commits/1e5d39d428bf713c878e3a4e391c954f6d37e651

gruble commented:
> Det går kjapt å hente observasjoner fra lokal database og virker som filter på kartutsnitt fungerer utmerket i listevisning.
> 
> Når jeg bytter språk er det litt som skjer:
> 
> 1. Jeg får denne i konsollet:
>    TypeError: Cannot read properties of undefined (reading 'sync_time_ms') at sqlite.service.ts:225:29
> 2. Observasjonsmarkører i kartet forsvinner, selv om det ser ut som den henter observasjoner på nytt fra API'et og lagrer dem i databasen. Markørene vises ikke når den er ferdig med synk.
> 3. Det tar rundt to minutter å hente og lagre alle observasjoner på nytt når man bytter språk (i prod nå). Mulig det tar mye tid pga. mye logging, men det kan kanskje være verdt å prøve å kjøre insert av flere observasjoner om gangen.

Virker som dette er fikset og syncing etter språkbytte går mye kjappere nå.

gruble reviewed (approved):
Synes dette funker bra nå. Elegant løst og godt dokumentert. Mye ryddigere enn det vi har idag. Søkefunksjonalitet mot lokal database er mye bedre, så jeg kan ikke skjønne annet enn at responstid og minneforbruk for lokale søk må være kraftig redusert. Rett og slett et imponerende stykke arbeide.
Kult at sqlite har ferdig opplegg for oppgradering av skjema, det er nyttig. 
Kommentarene som står igjen er mest flisespikkeri. Jeg mener vi fint kan ta denne inn i develop, slik den er nå.
Om vi ønsker å bruke sql-pluginen til flere ting enn observasjoner på sikt, kan vi heller skille observasjonsspesifikke funksjoner og generelle databasefunksjoner i hver sine servicer når det blir aktuelt.
Jeg har laget en ekstra oppgave på å skjule filtre som ikke støttes: https://nveprojects.atlassian.net/browse/RO-2231

jorgkv committed:
Use SearchService to fetch reg for detail page

jorgkv committed:
Update lock file

jorgkv committed:
Bruk entall ikke flertall på tabellnavn

jorgkv committed:
Cleanup

jorgkv committed:
Fix table and db names

jorgkv committed:
Do not show toast before start wizard is completed
